subprojects/messaging/src/test/groovy/org/gradle/internal/event/DefaultListenerManagerTest.groovy

```diff
@@ -888,6 +890,28 @@ class DefaultListenerManagerTest extends ConcurrentSpec {
         instant.removed < instant.handled
     }
 
+    @Issue('https://github.com/gradle/gradle-private/issues/1031')
+    def concurrentAccessIsSafe() {
+        when:
+        List<Class> manyClasses = ClassPath.from(Thread.currentThread().getContextClassLoader()).getTopLevelClassesRecursive('java.util')*.load().findAll { it.isInterface() }
+
+        start {
+            manyClasses.each {
+                manager.getBroadcaster(it)
+            }
+        }
+        start {
+            100.times {
+                Thread.sleep(1)
```

Comment:
Does this test fail consistently without the fix and such a low `sleep(...)` value?

Reply:
No, if there're no `sleep`, in most cases the test won't fail. The reason is that `getBroadcaster()` is much more expensive than `addListener` so we need to make sure the second thread lasts a bit longer. This is a bit magic, I've used a bigger loop size:

```
            10000.times {
                manager.addListener(new Object())
                manager.removeListener(new Object())
            }
```